### ISSUE TITLE
Submit & Redirect not working as expected (3.0.5)

### ISSUE THREAD

**Issue by yatiac** [getgrav/grav-plugin-form#355]
The submit action is not working as expected when there is no explicit redirection, the base url was removed in v3.0.5 and this is causing an issue when just submiting but no redirecting. (it was working fine in v3.0.4)

**Comment by rhukster** [comment]
If you have a `reset: true` make sure it's **NOT** after the `redirect:` in the form actions.. it will clear the redirect out.

**Comment by yatiac** [comment]
this is my form:

```
title: Contacto
form:
    name: contact
    fields:
        name:
            label: Nombre
            placeholder: 'Introduzca su nombre'
            autocomplete: 'off'
            type: text
            validate:
                required: true
        email:
            label: Email
            placeholder: 'Introduzca su e-mail'
            type: email
            validate:
                required: true
        message:
            label: Mensaje
            placeholder: 'Introduzca su mensaje'
            type: textarea
            validate:
                required: true
    buttons:
        submit:
            type: submit
            value: Enviar
            classes: button
    process:
        email:
            subject: '[Contacto desde Web] {{ form.value.name|e }}'
            body: '{% include ''forms/data.html.twig'' %}'
        message: 'Gracias por su interés, estaremos contactándolo en breve'
```

and it was working in v3.0.4

**Comment by rhukster** [comment]
Dropped your form into my test site.  Worked fine:

<img width="1114" alt="Image 2019-06-25 at 3 06 39 PM" src="https://user-images.githubusercontent.com/1084697/60133585-dea95180-975a-11e9-9d38-b7b583fc184e.png">

**Comment by yatiac** [comment]
is your website in the root folder of your webserver? I have mine inside a 'grav' folder and when redirected it is removing the base url (a.k.a 'grav' folder in the url)

**Comment by rhukster** [comment]
Nope, actually in a subfolder.  The thing is that it should not be redirecting at all, it should be submitting to the page the form sits on, then the message should show up on that form on when successfully submitted.

**Comment by yatiac** [comment]
For some reason it is redirecting for me.  

I think it could be related to this commit: https://github.com/getgrav/grav-plugin-form/commit/f917a1840284e17ac7801edc3e6fba35d8114e52

**Comment by rhukster** [comment]
Try manually reverting that fix, and then you will know for sure.

**Comment by rhukster** [comment]
Also inspect your form before you submit, Is the action wrong there? what does it look like?

<img width="751" alt="Image 2019-06-25 at 3 18 13 PM" src="https://user-images.githubusercontent.com/1084697/60134309-778c9c80-975c-11e9-8891-3d058e753e31.png">

**Comment by rhukster** [comment]
Do you have a `custom_base_url:` set in system.yaml or something?

**Comment by yatiac** [comment]
the `action` seems to be wrong:
![image](https://user-images.githubusercontent.com/1914303/60134357-f685d300-9764-11e9-8378-caf97eff0c5c.png)
I manually changed it to `/grav/contacto` and it worked fine

**Comment by yatiac** [comment]
also, I dont have `custom_base_url:` set 
![image](https://user-images.githubusercontent.com/1914303/60134504-3b116e80-9765-11e9-871b-cd69af80b7e5.png)

**Comment by rhukster** [comment]
So did reverting that fix, make things work as expected?

**Comment by yatiac** [comment]
Yes, it did

**Comment by rhukster** [comment]
ok try this instead:

```
{% set action = form.action ? url(form.action) : page.rawRoute() ~ uri.params %}
```

**Comment by yatiac** [comment]
that didn't work

**Comment by rhukster** [comment]
yah its not falling back, it's the action in the form object, just not sure why yours is not working when it does on mine and similar setup.

**Comment by rhukster** [comment]
BTW what version of Grav are you running?

**Comment by yatiac** [comment]
1.6.11

**Comment by rhukster** [comment]
Anything else you can tell me about the your setup? webserver? platform? any vhosts? anything special in how you configured things? I just can't replicate your issue :(

**Comment by yatiac** [comment]
I don't have anything special at all, the site is installed in a `grav` folder in an ubuntu server runing php 7.2

interesting thing, it was working in v3.0.4 and reverting the commit makes it work again so it is confirmed is https://github.com/getgrav/grav-plugin-form/commit/f917a1840284e17ac7801edc3e6fba35d8114e52

**Comment by rhukster** [comment]
yah, but the thing is that its now using a 'smarter' function in URL which should append the 'root', ie `grav/` in your case in front of the page route.  This is what it does for me.  The issue is the way we had it before, it inserted the root no matter what the route was, so there were scenarios where the `grav/` root was already in the URL, so you ended up with double: `grav/grav/contacto` for example.  It's smarter now to only add it if it needs it....  but for you its not adding it at all...

Can you turn on your debugger and add this in the `user/plugins/form/templates/forms/default/form.html.twig` file under the set action:

```
{{ dump(uri.rootUrl) }}
```

It should output `/grav` in the messages panel

**Comment by yatiac** [comment]
it does

**Comment by rhukster** [comment]
so if you put this: what is the output:

```
{{ dump url('/contacto') }}
```

??

**Comment by yatiac** [comment]
![image](https://user-images.githubusercontent.com/1914303/60136232-3cdd3100-9769-11e9-98a4-35baf67fa3f4.png)

**Comment by rhukster** [comment]
see my edit.. forgot to wrap it in () the first time.

**Comment by yatiac** [comment]
sorry, I missed it, it outputs `/contacto`

**Comment by rhukster** [comment]
Well this makes no sense... We use this `url()` twig function all over the place, and never had any reports that its unreliable, or returns wrong results. 

Can you please ZIP up your entire site and send to me at rhuk at getgrav dot org?

**Comment by yatiac** [comment]
Is there anything else I can send you instead of the whole site?

Also, thank you so much for your help troubleshooting this, right now it is working with the manual revertion.

**Comment by rhukster** [comment]
to really recreate your setup, i need your whole site.  if you want to remove any other pages other than the ones in question that's fine, they won't impact. but need config/plugins/overrides etc

**Comment by yatiac** [comment]
My email client is blocking it.

**Comment by rhukster** [comment]
Drop it on dropbox/gdrive/onedrive/etc and send me a link?

**Comment by yatiac** [comment]
Done

**Comment by rhukster** [comment]
Thanks that helped a tonne. I figured out the issue as soon as I loaded your page. So the problem is that the Grav `url()` twig function is handling things correctly, but you are using a Gantry theme, and that has it's own `url()` function that is not taking the site's base_url [$uri->rootUrl()] into account.

This is actually a bigger Gantry on Grav issue that we need to get resolved. So i'm going to ping @mahagr about this as it will undoubtedly cause other problems with Grav-based Gantry themes where the `url()` function is used.

BTW, I had JS errors in admin and had to reinstall the admin plugin via direct-install to resolve.  I suspect you are uploading files via FTP or something and that is not very reliable. Better to upload the ZIP and use a local file manager UI to extract the files. Or better yet, use SSH and GPM in CLI, or GPM via the admin.

**Comment by mahagr** [comment]
Umm, it looks like that in Gantry `url()` assumes all the URLs without registered stream as being external. This may be tricky to fix as many sites may rely on that behaviour.

**Comment by mahagr** [comment]
I'm going to override the default behaviour in Grav only for now.

**Comment by mahagr** [comment]
@rhukster It looks like that Gantry has better support for URLs outside of the sites than Grav -- Grav always assumes that all the paths are in the Grav site, Gantry supports `/foo.jpg` (http://domain.com/foo.jpg), `foo.jpg` (http://domain.com/grav/foo.jpg) as well as '../foo.jpg` (http://domain.com/foo.jpg).

All of this said, we really need to stop using `url()` for routes -- it'll not work with multi-lang for example.

**Comment by yatiac** [comment]
Thank you for all your help!

**Comment by virusx** [comment]
I had an issue after updating from Form 3.0.1 to 3.0.7.
I checked with WinMerge the difference between the update and the previous plugin.
I tracked down the error and it was caused by the update in "user\plugins\form\templates\forms\default\form.html.twig"
from `{% set action = form.action ? base_url ~ form.action : base_url ~ page.route ~ uri.params %}` to `{% set action = url(form.action ?: page.route ~ uri.params) %}`
As soon as I copied back this line, the plugin worked and I found this thread.
I'm using Ceevee template with all last update in GRAV and plugins.
